src/hpke.rs

```diff
@@ -0,0 +1,1072 @@
+    type Error = error::Unspecified;
+
+    fn try_from(v: u16) -> Result<Self, Self::Error> {
+        match v {
+            0x0001 if 0x0001 == KdfId::Sha256 as u16 => Ok(KdfId::Sha256),
+            0x0002 if 0x0002 == KdfId::Sha384 as u16 => Ok(KdfId::Sha384),
+            0x0003 if 0x0003 == KdfId::Sha512 as u16 => Ok(KdfId::Sha512),
+            _ => Err(error::Unspecified),
+        }
+    }
+}
+
+/// AeadId is the Authenticated Encryption with Associated Data algorithm identifier.
+/// Aead is used for encrypting messages.
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[non_exhaustive]
+#[repr(u16)]
+pub enum AeadId {
+    /// AES-128-GCM
+    Aes128Gcm = 0x0001,
+    /// AES-256-GCM
+    Aes256Gcm = 0x0002,
+    /// ChaCha20Poly1305
+    ChaCha20Poly1305 = 0x0003,
+    // TODO: Implement ExportOnly
+}
+
+impl TryFrom<u16> for AeadId {
+    type Error = error::Unspecified;
+
+    fn try_from(v: u16) -> Result<Self, Self::Error> {
+        match v {
+            0x0001 if 0x0001 == AeadId::Aes128Gcm as u16 => Ok(AeadId::Aes128Gcm),
+            0x0002 if 0x0002 == AeadId::Aes256Gcm as u16 => Ok(AeadId::Aes256Gcm),
+            0x0003 if 0x0003 == AeadId::ChaCha20Poly1305 as u16 => Ok(AeadId::ChaCha20Poly1305),
+            _ => Err(error::Unspecified),
+        }
+    }
+}
+
+/// KeyPair
+pub struct KeyPair {
+    secret_key: ec::Seed,
+    public_key: [u8; MAX_PUBLIC_KEY_LEN],
+    /// public key length
+    pub public_key_len: u16,
+    /// secret key length
+    pub secret_key_len: u16,
+}
+
+impl KeyPair {
+    /// Returns the public key as a byte slice
+    pub fn public_key_bytes(&self) -> &[u8] {
+        &self.public_key[..self.public_key_len as usize]
+    }
+}
+
+/// Suite
+pub struct Suite {
+    suite_id_context: [u8; 10],
+    hash: &'static hkdf::Algorithm,
+    aead: &'static aead::Algorithm,
+    kem: Kem,
+}
+
+fn get_suite_id_kem(kem_id: KemId) -> [u8; 5] {
+    let mut kem = [0u8; 5];
+    kem[0..3].copy_from_slice("KEM".as_bytes());
+    kem[3..5].copy_from_slice(&(kem_id as u16).to_be_bytes());
+    kem
+}
+
+fn get_suite_id_context(kem_id: KemId, kdf_id: KdfId, aead_id: AeadId) -> [u8; 10] {
+    let mut suite = [0u8; 10];
+    suite[..4].copy_from_slice("HPKE".as_bytes());
+    let mut i = 4;
+    for id in [kem_id as u16, kdf_id as u16, aead_id as u16] {
+        suite[i..i + 2].copy_from_slice(&id.to_be_bytes());
+        i += 2;
+    }
+    suite
+}
+
+/// Kem
+pub struct Kem {
+    suite_id_kem: [u8; 5],
+    kem_id: KemId,
+    hash: &'static hkdf::Algorithm,
+    agreement: &'static agreement::Algorithm,
+}
+
+impl Kem {
+    /// Instantiate a new KEM struct from a given [KemId]
+    pub fn new(kem_id: KemId) -> Result<Self, error::Unspecified> {
+        match kem_id {
+            KemId::P256Sha256 => Ok(Kem {
+                suite_id_kem: get_suite_id_kem(kem_id),
+                kem_id,
+                hash: &hkdf::HKDF_SHA256,
+                agreement: &agreement::ECDH_P256,
```

Comment:
The unusual structure of *ring*, e.g. in how we have "HMAC algorithms" that are thin wrappers around digest algorithms, is to facilitate link-time dead-code elimination by ensuring that we never do this kind of dispatching that potentially would pull in every agreement::Algorithm algorithm.

Following that, it would be better if we had something like:
```
struct Algorithm {
    hkdf: &'static hkdf::Algorithm,
    agreement: &'static agreement::Algorithm
} 
static P256_SHA256: Algorithm = Algorithm {
    hkdf: hkdf::HKDF_SHA56,
    agreement: agreement::ECDH_P256
};
```
In the case of the unimplemented ones, we can just omit them until everything else is OK.

When we need to convert a on-the-wire identifier to an `Algorithm`, we need a function like this:

```
fn algorithm_by_id(
    allowed_algorithms: &[&'static Algorithm], 
    on_the_wire_id: u16)
    -> Result<'static Algorithm, error::Unspecified>
{
    allowed_algorithms.iter().find(|alg| alg.id == on_the_wire_id).ok_or(error::Unspecified)
}
```

This way, we again avoid statically referencing any algorithms, and then allow the linker to discard unused ones.

At the top level of the API, we need the caller to be able to pass in a list (slice/iterator) of allowed algorithms. This is the mechanism that lets the calling code (e.g. Rustls) choose which algorithms it wants to allow the use of, and also the priority. We usually use slices because we try to avoid allocations, and that tends to work OK (AFAICT) because the number of different algorithms we support is low so linear search through them is OK.